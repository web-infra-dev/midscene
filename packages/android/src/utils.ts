import { getDebug } from '@midscene/shared/logger';
import { ADB, type Device } from 'appium-adb';

const debugUtils = getDebug('android:utils');

export async function getConnectedDevices(): Promise<Device[]> {
  try {
    const adb = await ADB.createADB({
      adbExecTimeout: 60000,
    });
    const devices = await adb.getConnectedDevices();

    debugUtils(`Found ${devices.length} connected devices: `, devices);

    return devices;
  } catch (error: any) {
    console.error('Failed to get device list:', error);
    throw new Error(
      `Unable to get connected Android device list, please check https://midscenejs.com/integrate-with-android.html#faq : ${error.message}`,
      {
        cause: error,
      },
    );
  }
}
