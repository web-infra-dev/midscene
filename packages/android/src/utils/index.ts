import { ADB, type Device } from 'appium-adb';
import { debugPage } from '../page';
export async function getConnectedDevices(): Promise<Device[]> {
  try {
    const adb = await ADB.createADB({
      adbExecTimeout: 60000,
    });
    const devices = await adb.getConnectedDevices();

    debugPage(`Found ${devices.length} connected devices: `, devices);

    return devices;
  } catch (error) {
    console.error('Failed to get device list:', error);
    throw new Error('Unable to get connected Android device list');
  }
}
