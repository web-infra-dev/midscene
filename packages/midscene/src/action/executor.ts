import assert from 'node:assert';
import type {
  ExecutionDump,
  ExecutionTask,
  ExecutionTaskApply,
  ExecutionTaskInsightLocateOutput,
  ExecutionTaskReturn,
  ExecutorContext,
} from '@/types';
import { getMidscenePkgInfo } from '@midscene/shared/fs';

export class Executor {
  name: string;

  description?: string;

  tasks: ExecutionTask[];

  // status of executor
  status: 'init' | 'pending' | 'running' | 'completed' | 'error';

  constructor(
    name: string,
    description?: string,
    tasks?: ExecutionTaskApply[],
  ) {
    this.status = tasks && tasks.length > 0 ? 'pending' : 'init';
    this.name = name;
    this.description = description;
    this.tasks = (tasks || []).map((item) => this.markTaskAsPending(item));
  }

  private markTaskAsPending(task: ExecutionTaskApply): ExecutionTask {
    return {
      status: 'pending',
      ...task,
    };
  }

  async append(task: ExecutionTaskApply[] | ExecutionTaskApply): Promise<void> {
    assert(
      this.status !== 'error',
      'executor is in error state, cannot append task',
    );
    if (Array.isArray(task)) {
      this.tasks.push(...task.map((item) => this.markTaskAsPending(item)));
    } else {
      this.tasks.push(this.markTaskAsPending(task));
    }
    if (this.status !== 'running') {
      this.status = 'pending';
    }
  }

  async flush(): Promise<any> {
    if (this.status === 'init' && this.tasks.length > 0) {
      console.warn(
        'illegal state for executor, status is init but tasks are not empty',
      );
    }

    assert(this.status !== 'running', 'executor is already running');
    assert(this.status !== 'completed', 'executor is already completed');
    assert(this.status !== 'error', 'executor is in error state');

    const nextPendingIndex = this.tasks.findIndex(
      (task) => task.status === 'pending',
    );
    if (nextPendingIndex < 0) {
      // all tasks are completed
      return;
    }

    this.status = 'running';
    let taskIndex = nextPendingIndex;
    let successfullyCompleted = true;

    let previousFindOutput: ExecutionTaskInsightLocateOutput | undefined;

    while (taskIndex < this.tasks.length) {
      const task = this.tasks[taskIndex];
      assert(
        task.status === 'pending',
        `task status should be pending, but got: ${task.status}`,
      );
      task.timing = {
        start: Date.now(),
      };
      try {
        task.status = 'running';
        assert(
          ['Insight', 'Action', 'Planning'].indexOf(task.type) >= 0,
          `unsupported task type: ${task.type}`,
        );

        const { executor, param } = task;
        assert(executor, `executor is required for task type: ${task.type}`);

        let returnValue;
        const executorContext: ExecutorContext = {
          task,
          element: previousFindOutput?.element,
        };
        if (task.type === 'Insight') {
          assert(
            task.subType === 'Locate' ||
              task.subType === 'Query' ||
              task.subType === 'Assert',
            `unsupported insight subType: ${task.subType}`,
          );
          returnValue = await task.executor(param, executorContext);
          if (task.subType === 'Locate') {
            previousFindOutput = (
              returnValue as ExecutionTaskReturn<ExecutionTaskInsightLocateOutput>
            )?.output;
          }
        } else if (task.type === 'Action' || task.type === 'Planning') {
          returnValue = await task.executor(param, executorContext);
        } else {
          console.warn(
            `unsupported task type: ${task.type}, will try to execute it directly`,
          );
          returnValue = await task.executor(param, executorContext);
        }

        Object.assign(task, returnValue);
        task.status = 'finished';
        task.timing.end = Date.now();
        task.timing.cost = task.timing.end - task.timing.start;
        taskIndex++;
      } catch (e: any) {
        successfullyCompleted = false;
        task.error = e?.message || 'error-without-message';
        task.errorStack = e.stack;

        task.status = 'failed';
        task.timing.end = Date.now();
        task.timing.cost = task.timing.end - task.timing.start;
        break;
      }
    }

    // set all remaining tasks as cancelled
    for (let i = taskIndex + 1; i < this.tasks.length; i++) {
      this.tasks[i].status = 'cancelled';
    }

    if (successfullyCompleted) {
      this.status = 'completed';
    } else {
      this.status = 'error';
    }
    if (this.tasks.length) {
      // return the last output
      const outputIndex = Math.min(taskIndex, this.tasks.length - 1);
      return this.tasks[outputIndex].output;
    }
  }

  isInErrorState(): boolean {
    return this.status === 'error';
  }

  latestErrorTask(): ExecutionTask | null {
    if (this.status !== 'error') {
      return null;
    }
    const errorTaskIndex = this.tasks.findIndex(
      (task) => task.status === 'failed',
    );
    if (errorTaskIndex >= 0) {
      return this.tasks[errorTaskIndex];
    }
    return null;
  }

  dump(): ExecutionDump {
    const dumpData: ExecutionDump = {
      sdkVersion: getMidscenePkgInfo(__dirname).version,
      logTime: Date.now(),
      name: this.name,
      description: this.description,
      tasks: this.tasks,
    };
    return dumpData;
  }
}
