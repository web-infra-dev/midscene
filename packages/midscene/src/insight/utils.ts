import assert from 'node:assert';
import { randomUUID } from 'node:crypto';
/* eslint-disable @typescript-eslint/ban-types */
import { existsSync } from 'node:fs';
import { join } from 'node:path';
import type {
  BaseElement,
  DumpMeta,
  DumpSubscriber,
  ElementById,
  InsightDump,
  LiteUISection,
  PartialInsightDumpFromSDK,
  Rect,
  UIContext,
  UISection,
} from '@/types';
import {
  getLogDir,
  insightDumpFileExt,
  stringifyDumpData,
  writeLogFile,
} from '@/utils';
import { getMidscenePkgInfo } from '@midscene/shared/fs';

let logFileName = '';
const logContent: string[] = [];
const logIdIndexMap: Record<string, number> = {};
const { pid } = process;
const logFileExt = insightDumpFileExt;

export function writeInsightDump(
  data: PartialInsightDumpFromSDK,
  logId?: string,
  dumpSubscriber?: DumpSubscriber,
): string {
  const logDir = getLogDir();
  assert(logDir, 'logDir should be set before writing dump file');

  const id = logId || randomUUID();
  const baseData: DumpMeta = {
    sdkVersion: getMidscenePkgInfo(__dirname).version,
    logTime: Date.now(),
  };
  const finalData: InsightDump = {
    logId: id,
    ...baseData,
    ...data,
  };

  dumpSubscriber?.(finalData);

  if (!logFileName) {
    logFileName = `pid_${pid}_${baseData.logTime}`;
    while (existsSync(join(logDir, `${logFileName}.${logFileExt}`))) {
      logFileName = `${pid}_${baseData.logTime}-${Math.random()}`;
    }
  }
  const dataString = stringifyDumpData(finalData, 2);

  if (typeof logIdIndexMap[id] === 'number') {
    logContent[logIdIndexMap[id]] = dataString;
  } else {
    const length = logContent.push(dataString);
    logIdIndexMap[id] = length - 1;
  }
  writeLogFile({
    fileName: logFileName,
    fileExt: logFileExt,
    fileContent: `[\n${logContent.join(',\n')}\n]`,
    type: 'dump',
  });

  return id;
}

export function idsIntoElements(
  ids: string[],
  elementById: ElementById,
): BaseElement[] {
  return ids.reduce<BaseElement[]>((acc, id) => {
    const element = elementById(id);
    if (element) {
      acc.push(element);
    } else {
      console.warn(`element not found by id: ${id}`);
    }
    return acc;
  }, []);
}

// biome-ignore lint/complexity/noBannedTypes: <explanation>
export function shallowExpandIds<DataScheme extends object = {}>(
  data: DataScheme,
  ifMeet: (id: string) => boolean,
  elementsById: (id: string) => BaseElement | BaseElement[] | null,

  // return same type as data
): DataScheme {
  const keys = Object.keys(data);
  keys.forEach((key) => {
    const value = (data as any)[key];
    if (typeof value === 'string' && ifMeet(value)) {
      // (data as any)[key] = elementsById(value);
      (data as any)[key] = elementsById(value);
    } else if (Array.isArray(value)) {
      const newValue = value.map((id) => (ifMeet(id) ? elementsById(id) : id));
      (data as any)[key] = newValue;
    }
  });

  return data;
}

export function expandLiteSection(
  liteSection: LiteUISection,
  elementById: ElementById,
): UISection {
  const { textIds, ...remainingFields } = liteSection;

  const texts: BaseElement[] = idsIntoElements(textIds, elementById);

  let leftMost = -1;
  let topMost = -1;
  let rightMost = -1;
  let bottomMost = -1;
  texts.forEach((text) => {
    leftMost =
      leftMost === -1 ? text.rect.left : Math.min(leftMost, text.rect.left);
    topMost = topMost === -1 ? text.rect.top : Math.min(topMost, text.rect.top);
    rightMost = Math.max(rightMost, text.rect.left + text.rect.width);
    bottomMost = Math.max(bottomMost, text.rect.top + text.rect.height);
  });
  const sectionRect: Rect = {
    left: leftMost,
    top: topMost,
    width: rightMost - leftMost,
    height: bottomMost - topMost,
  };

  const section: UISection = {
    ...remainingFields,
    content: texts,
    rect: sectionRect,
  };

  return section;
}

/**
 * a fake parser that collects all text into a single section
 * It's useful for debugging and testing
 * @param context
 * @returns
 */
export async function fakeParserCollectsTexts<P>(context: UIContext): Promise<{
  [K in keyof P]: UISection;
}> {
  const { content } = context;
  const section: UISection = {
    name: 'all-texts',
    description: 'all texts in the page',
    sectionCharacteristics: 'all texts',
    content,
    rect: {
      left: 0,
      top: 0,
      width: context.size.width,
      height: context.size.height,
    },
  };
  return {
    'all-texts': section,
  } as any;
}
