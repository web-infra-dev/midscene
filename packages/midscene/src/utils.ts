import assert from 'node:assert';
import { randomUUID } from 'node:crypto';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { basename, dirname, join } from 'node:path';
import { getMidscenePkgInfo } from '@midscene/shared/fs';
import type { Rect, ReportDumpWithAttributes } from './types';

interface PkgInfo {
  name: string;
  version: string;
  dir: string;
}

const midscenePkgInfo = getMidscenePkgInfo(__dirname);
let logDir = join(process.cwd(), './midscene_run/');
let logEnvReady = false;
export const insightDumpFileExt = 'insight-dump.json';
export const groupedActionDumpFileExt = 'web-dump.json';

export function getLogDir() {
  return logDir;
}

export function setLogDir(dir: string) {
  logDir = dir;
}

export function getLogDirByType(type: 'dump' | 'cache' | 'report') {
  const dir = join(getLogDir(), type);
  if (!existsSync(dir)) {
    mkdirSync(dir, { recursive: true });
  }
  return dir;
}

export function writeDumpReport(
  fileName: string,
  dumpData: string | ReportDumpWithAttributes[],
) {
  const { dir } = midscenePkgInfo;
  const reportTplPath = join(dir, './report/index.html');
  existsSync(reportTplPath) ||
    assert(false, `report template not found: ${reportTplPath}`);
  const reportPath = join(getLogDirByType('report'), `${fileName}.html`);
  const tpl = readFileSync(reportTplPath, 'utf-8');
  let reportContent: string;
  if (
    (Array.isArray(dumpData) && dumpData.length === 0) ||
    typeof dumpData === 'undefined'
  ) {
    reportContent = tpl.replace(
      '{{dump}}',
      '<script type="midscene_web_dump" type="application/json"></script>',
    );
  } else if (typeof dumpData === 'string') {
    reportContent = tpl.replace(
      '{{dump}}',
      `<script type="midscene_web_dump" type="application/json">${dumpData}</script>`,
    );
  } else {
    const dumps = dumpData.map(({ dumpString, attributes }) => {
      const attributesArr = Object.keys(attributes || {}).map((key) => {
        return `${key}="${encodeURIComponent(attributes![key])}"`;
      });
      return `<script type="midscene_web_dump" type="application/json" ${attributesArr.join(' ')}>${dumpString}</script>`;
    });
    reportContent = tpl.replace('{{dump}}', dumps.join('\n'));
  }
  writeFileSync(reportPath, reportContent);

  return reportPath;
}

export function writeLogFile(opts: {
  fileName: string;
  fileExt: string;
  fileContent: string;
  type: 'dump' | 'cache' | 'report';
  generateReport?: boolean;
}) {
  const { fileName, fileExt, fileContent, type = 'dump' } = opts;
  const targetDir = getLogDirByType(type);
  // Ensure directory exists
  if (!logEnvReady) {
    assert(targetDir, 'logDir should be set before writing dump file');

    // gitIgnore in the parent directory
    const gitIgnorePath = join(targetDir, '../../.gitignore');
    let gitIgnoreContent = '';
    if (existsSync(gitIgnorePath)) {
      gitIgnoreContent = readFileSync(gitIgnorePath, 'utf-8');
    }

    // ignore the log folder
    const logDirName = basename(logDir);
    if (!gitIgnoreContent.includes(`${logDirName}/`)) {
      writeFileSync(
        gitIgnorePath,
        `${gitIgnoreContent}\n# Midscene.js dump files\n${logDirName}/report\n${logDirName}/dump\n`,
        'utf-8',
      );
    }
    logEnvReady = true;
  }

  const filePath = join(targetDir, `${fileName}.${fileExt}`);

  const outputResourceDir = dirname(filePath);
  if (!existsSync(outputResourceDir)) {
    mkdirSync(outputResourceDir, { recursive: true });
  }

  writeFileSync(filePath, fileContent);

  if (opts?.generateReport) {
    return writeDumpReport(fileName, fileContent);
  }

  return filePath;
}

export function getTmpDir() {
  const path = join(tmpdir(), midscenePkgInfo.name);
  mkdirSync(path, { recursive: true });
  return path;
}

export function getTmpFile(fileExtWithoutDot: string) {
  const filename = `${randomUUID()}.${fileExtWithoutDot}`;
  return join(getTmpDir(), filename);
}

export function overlapped(container: Rect, target: Rect) {
  // container and the target have some part overlapped
  return (
    container.left < target.left + target.width &&
    container.left + container.width > target.left &&
    container.top < target.top + target.height &&
    container.top + container.height > target.top
  );
}

export async function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export function replacerForPageObject(key: string, value: any) {
  if (value && value.constructor?.name === 'Page') {
    return '[Page object]';
  }
  if (value && value.constructor?.name === 'Browser') {
    return '[Browser object]';
  }
  return value;
}

export function stringifyDumpData(data: any, indents?: number) {
  return JSON.stringify(data, replacerForPageObject, indents);
}
