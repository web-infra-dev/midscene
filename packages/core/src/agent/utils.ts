import { elementByPositionWithElementInfo } from '@/ai-model';
import type { AbstractInterface } from '@/device';
import type {
  BaseElement,
  ElementTreeNode,
  ExecutionDump,
  ExecutionTask,
  ExecutorContext,
  PlanningLocateParam,
  TMultimodalPrompt,
  TUserPrompt,
  UIContext,
} from '@/index';
import { uploadTestInfoToServer } from '@/utils';
import {
  MIDSCENE_REPORT_TAG_NAME,
  globalConfigManager,
} from '@midscene/shared/env';
import {
  generateElementByPosition,
  getNodeFromCacheList,
} from '@midscene/shared/extractor';
import { resizeImgBase64 } from '@midscene/shared/img';
import { getDebug } from '@midscene/shared/logger';
import { _keyDefinitions } from '@midscene/shared/us-keyboard-layout';
import { assert, logMsg, uuid } from '@midscene/shared/utils';
import dayjs from 'dayjs';
import { debug as cacheDebug } from './task-cache';
import type { TaskExecutor } from './tasks';

const debugProfile = getDebug('web:tool:profile');

export async function commonContextParser(
  interfaceInstance: AbstractInterface,
): Promise<UIContext> {
  assert(interfaceInstance, 'interfaceInstance is required');

  debugProfile('Getting interface description');
  const description = interfaceInstance.describe?.() || '';
  debugProfile('Interface description end');

  debugProfile('Uploading test info to server');
  uploadTestInfoToServer({ testUrl: description });
  debugProfile('UploadTestInfoToServer end');

  let screenshotBase64 = await interfaceInstance.screenshotBase64();
  assert(screenshotBase64!, 'screenshotBase64 is required');

  const size = await interfaceInstance.size();
  debugProfile(`size: ${size.width}x${size.height} dpr: ${size.dpr}`);

  if (size.dpr && size.dpr > 1) {
    debugProfile('Resizing screenshot for high DPR display');
    screenshotBase64 = await resizeImgBase64(screenshotBase64, {
      width: size.width,
      height: size.height,
    });
    debugProfile('ResizeImgBase64 end');
  }

  return {
    tree: {
      node: null,
      children: [],
    },
    size,
    screenshotBase64: screenshotBase64!,
  };
}

export function getReportFileName(tag = 'web') {
  const reportTagName = globalConfigManager.getEnvConfigValue(
    MIDSCENE_REPORT_TAG_NAME,
  );
  const dateTimeInFileName = dayjs().format('YYYY-MM-DD_HH-mm-ss');
  // ensure uniqueness at the same time
  const uniqueId = uuid().substring(0, 8);
  return `${reportTagName || tag}-${dateTimeInFileName}-${uniqueId}`;
}

export function printReportMsg(filepath: string) {
  logMsg(`Midscene - report file updated: ${filepath}`);
}

/**
 * Get the current execution file name
 * @returns The name of the current execution file
 */
export function getCurrentExecutionFile(trace?: string): string | false {
  const error = new Error();
  const stackTrace = trace || error.stack;
  const pkgDir = process.cwd() || '';
  if (stackTrace) {
    const stackLines = stackTrace.split('\n');
    for (const line of stackLines) {
      if (
        line.includes('.spec.') ||
        line.includes('.test.') ||
        line.includes('.ts') ||
        line.includes('.js')
      ) {
        const match = line.match(/(?:at\s+)?(.*?\.(?:spec|test)\.[jt]s)/);
        if (match?.[1]) {
          const targetFileName = match[1]
            .replace(pkgDir, '')
            .trim()
            .replace('at ', '');
          return targetFileName;
        }
      }
    }
  }
  return false;
}

const testFileIndex = new Map<string, number>();

export function generateCacheId(fileName?: string): string {
  let taskFile = fileName || getCurrentExecutionFile();
  if (!taskFile) {
    taskFile = uuid();
    console.warn(
      'Midscene - using random UUID for cache id. Cache may be invalid.',
    );
  }

  if (testFileIndex.has(taskFile)) {
    const currentIndex = testFileIndex.get(taskFile);
    if (currentIndex !== undefined) {
      testFileIndex.set(taskFile, currentIndex + 1);
    }
  } else {
    testFileIndex.set(taskFile, 1);
  }
  return `${taskFile}-${testFileIndex.get(taskFile)}`;
}

export function matchElementFromPlan(
  planLocateParam: PlanningLocateParam,
  tree: ElementTreeNode<BaseElement>,
) {
  if (!planLocateParam) {
    return undefined;
  }
  if (planLocateParam.id) {
    return getNodeFromCacheList(planLocateParam.id);
  }

  if (planLocateParam.bbox) {
    const centerPosition = {
      x: Math.floor((planLocateParam.bbox[0] + planLocateParam.bbox[2]) / 2),
      y: Math.floor((planLocateParam.bbox[1] + planLocateParam.bbox[3]) / 2),
    };
    let element = elementByPositionWithElementInfo(tree, centerPosition);

    if (!element) {
      element = generateElementByPosition(centerPosition) as BaseElement;
    }

    return element;
  }

  return undefined;
}

export async function matchElementFromCache(
  taskExecutor: TaskExecutor,
  xpaths: string[] | undefined,
  cachePrompt: TUserPrompt,
  cacheable: boolean | undefined,
) {
  try {
    if (
      xpaths?.length &&
      taskExecutor.taskCache?.isCacheResultUsed &&
      cacheable !== false &&
      (taskExecutor.interface as any).getElementInfoByXpath
    ) {
      // hit cache, use new id
      for (let i = 0; i < xpaths.length; i++) {
        const element = await (
          taskExecutor.interface as any
        ).getElementInfoByXpath(xpaths[i]);

        if (element?.id) {
          cacheDebug('cache hit, prompt: %s', cachePrompt);
          cacheDebug(
            'found a new element with same xpath, xpath: %s, id: %s',
            xpaths[i],
            element?.id,
          );
          return element;
        }
      }
    }
  } catch (error) {
    cacheDebug('get element info by xpath error: ', error);
  }
}

export function trimContextByViewport(execution: ExecutionDump) {
  function filterVisibleTree(
    node: ElementTreeNode<BaseElement>,
  ): ElementTreeNode<BaseElement> | null {
    if (!node) return null;

    // recursively process all children
    const filteredChildren = Array.isArray(node.children)
      ? (node.children
          .map(filterVisibleTree)
          .filter((child) => child !== null) as ElementTreeNode<BaseElement>[])
      : [];

    // if the current node is visible, keep it and the filtered children
    if (node.node && node.node.isVisible === true) {
      return {
        ...node,
        children: filteredChildren,
      };
    }

    // if the current node is invisible, but has visible children, create an empty node to include these children
    if (filteredChildren.length > 0) {
      return {
        node: null,
        children: filteredChildren,
      };
    }

    // if the current node is invisible and has no visible children, return null
    return null;
  }

  return {
    ...execution,
    tasks: Array.isArray(execution.tasks)
      ? execution.tasks.map((task: ExecutionTask) => {
          const newTask = { ...task };
          if (task.uiContext?.tree) {
            newTask.uiContext = {
              ...task.uiContext,
              tree: filterVisibleTree(task.uiContext.tree) || {
                node: null,
                children: [],
              },
            };
          }
          return newTask;
        })
      : execution.tasks,
  };
}

declare const __VERSION__: string | undefined;

export const getMidsceneVersion = (): string => {
  if (typeof __VERSION__ !== 'undefined') {
    return __VERSION__;
  } else if (
    process.env.__VERSION__ &&
    process.env.__VERSION__ !== 'undefined'
  ) {
    return process.env.__VERSION__;
  }
  throw new Error('__VERSION__ inject failed during build');
};

export const parsePrompt = (
  prompt: TUserPrompt,
): {
  textPrompt: string;
  multimodalPrompt?: TMultimodalPrompt;
} => {
  if (typeof prompt === 'string') {
    return {
      textPrompt: prompt,
      multimodalPrompt: undefined,
    };
  }
  return {
    textPrompt: prompt.prompt,
    multimodalPrompt: prompt.images
      ? {
          images: prompt.images,
          convertHttpImage2Base64: !!prompt.convertHttpImage2Base64,
        }
      : undefined,
  };
};
