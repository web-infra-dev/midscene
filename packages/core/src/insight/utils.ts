import {
  MIDSCENE_MODEL_NAME,
  MIDSCENE_USE_QWEN_VL,
  MIDSCENE_USE_VLM_UI_TARS,
  getAIConfig,
  getAIConfigInBoolean,
} from '@/env';
import type {
  DumpMeta,
  DumpSubscriber,
  InsightDump,
  PartialInsightDumpFromSDK,
} from '@/types';
import { getLogDir, getVersion, stringifyDumpData } from '@/utils';
import { assert } from '@midscene/shared/utils';
import { uuid } from '@midscene/shared/utils';
const logContent: string[] = [];
const logIdIndexMap: Record<string, number> = {};

export function emitInsightDump(
  data: PartialInsightDumpFromSDK,
  logId?: string,
  dumpSubscriber?: DumpSubscriber,
): string {
  const logDir = getLogDir();
  assert(logDir, 'logDir should be set before writing dump file');

  const id = logId || uuid();
  let modelDescription = '';
  if (getAIConfigInBoolean(MIDSCENE_USE_VLM_UI_TARS)) {
    modelDescription = 'vlm-ui-tars mode';
  } else if (getAIConfigInBoolean(MIDSCENE_USE_QWEN_VL)) {
    modelDescription = 'qwen-vl mode';
  }

  const baseData: DumpMeta = {
    sdkVersion: getVersion(),
    logTime: Date.now(),
    model_name: getAIConfig(MIDSCENE_MODEL_NAME) || '',
    model_description: modelDescription,
  };
  const finalData: InsightDump = {
    logId: id,
    ...baseData,
    ...data,
  };

  dumpSubscriber?.(finalData);

  const dataString = stringifyDumpData(finalData, 2);

  if (typeof logIdIndexMap[id] === 'number') {
    logContent[logIdIndexMap[id]] = dataString;
  } else {
    const length = logContent.push(dataString);
    logIdIndexMap[id] = length - 1;
  }

  return id;
}
