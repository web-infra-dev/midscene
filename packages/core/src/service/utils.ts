import type { DumpMeta, PartialServiceDumpFromSDK, ServiceDump } from '@/types';
import { uuid } from '@midscene/shared/utils';

export function createServiceDump(
  data: PartialServiceDumpFromSDK,
): ServiceDump {
  const baseData: DumpMeta = {
    logTime: Date.now(),
  };
  const finalData: ServiceDump = {
    logId: uuid(),
    ...baseData,
    ...data,
  };

  return finalData;
}
