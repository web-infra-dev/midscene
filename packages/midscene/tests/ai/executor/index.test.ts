import { Executor } from '@/action/executor';
import type {
  DumpSubscriber,
  ExecutionTaskActionApply,
  ExecutionTaskInsightLocate,
  ExecutionTaskInsightLocateApply,
  InsightDump,
} from '@/index';
import { fakeInsight } from 'tests/utils';
import { describe, expect, it, vi } from 'vitest';

const insightFindTask = (shouldThrow?: boolean) => {
  let insightDump: InsightDump | undefined;
  const dumpCollector: DumpSubscriber = (dump) => {
    insightDump = dump;
  };
  const insight = fakeInsight('test-executor');
  insight.onceDumpUpdatedFn = dumpCollector;

  const insightFindTask: ExecutionTaskInsightLocateApply = {
    type: 'Insight',
    subType: 'Locate',
    param: {
      prompt: 'test',
    },
    async executor(param, taskContext) {
      if (shouldThrow) {
        const { task } = taskContext;
        task.output = 'error-output';
        await new Promise((resolve) => setTimeout(resolve, 100));
        throw new Error('test-error');
      }
      return {
        output: {
          element: await insight.locate(param.prompt),
        },
        log: {
          dump: insightDump,
        },
        cache: {
          hit: false,
        },
      };
    },
  };
  return insightFindTask;
};

vi.setConfig({
  testTimeout: 40 * 1000,
});

describe('executor', () => {
  it(
    'insight - basic run',
    async () => {
      const insightTask1 = insightFindTask();
      const flushResultData = 'abcdef';
      const taskParam = {
        action: 'tap',
        anything: 'acceptable',
      };
      const tapperFn = vi.fn();
      const actionTask: ExecutionTaskActionApply = {
        type: 'Action',
        param: taskParam,
        executor: tapperFn,
      };
      const actionTask2: ExecutionTaskActionApply = {
        type: 'Action',
        param: taskParam,
        executor: async () => {
          return {
            output: flushResultData,
          } as any;
        },
      };

      const inputTasks = [insightTask1, actionTask, actionTask2];

      const executor = new Executor(
        'test',
        'hello, this is a test',
        inputTasks,
      );
      const flushResult = await executor.flush();
      const tasks = executor.tasks as ExecutionTaskInsightLocate[];
      const { element } = tasks[0].output || {};
      expect(element).toBeTruthy();

      expect(tasks.length).toBe(inputTasks.length);
      expect(tasks[0].status).toBe('finished');
      expect(tasks[0].output).toMatchSnapshot();
      expect(tasks[0].log?.dump).toBeTruthy();
      expect(tasks[0].timing?.end).toBeTruthy();
      expect(tasks[0].cache).toBeTruthy();
      expect(tasks[0].cache?.hit).toEqual(false);

      expect(tapperFn).toBeCalledTimes(1);
      expect(tapperFn.mock.calls[0][0]).toBe(taskParam);
      expect(tapperFn.mock.calls[0][1].element).toBe(element);
      expect(tapperFn.mock.calls[0][1].task).toBeTruthy();

      const dump = executor.dump();
      expect(dump.logTime).toBeTruthy();

      expect(flushResult).toBe(flushResultData);
    },
    {
      timeout: 999 * 1000,
    },
  );

  it('insight - init and append', async () => {
    const initExecutor = new Executor('test');
    expect(initExecutor.status).toBe('init');
    const tapperFn = vi.fn();

    const insightTask1 = insightFindTask();
    const actionTask: ExecutionTaskActionApply = {
      type: 'Action',
      param: {
        action: 'tap',
        element: 'previous',
      },
      executor: async () => {
        // delay 500
        await new Promise((resolve) => setTimeout(resolve, 500));
        tapperFn();
      },
    };

    initExecutor.append(insightTask1);
    initExecutor.append(actionTask);
    expect(initExecutor.status).toBe('pending');
    expect(initExecutor.tasks.length).toBe(2);
    expect(tapperFn).toBeCalledTimes(0);

    const dumpContent1 = initExecutor.dump();
    expect(dumpContent1.tasks.length).toBe(2);

    // append while running
    const output = await Promise.all([
      initExecutor.flush(),
      (async () => {
        // sleep 200ms
        expect(initExecutor.status).toBe('running');
        await new Promise((resolve) => setTimeout(resolve, 200));
        initExecutor.append(actionTask);
        expect(initExecutor.status).toBe('running');
      })(),
    ]);

    expect(initExecutor.status).toBe('completed');
    expect(initExecutor.tasks.length).toBe(3);
    expect(initExecutor.tasks[2].status).toBe('finished');

    // append while completed
    initExecutor.append(actionTask);
    expect(initExecutor.status).toBe('pending');

    // same dumpPath to append
    const dumpContent2 = initExecutor.dump();
    expect(dumpContent2.tasks.length).toBe(4);

    expect(initExecutor.latestErrorTask()).toBeFalsy();
  });

  it('insight - run with error', async () => {
    const executor = new Executor('test', 'test-description', [
      insightFindTask(true),
      insightFindTask(),
    ]);
    const r = await executor.flush();
    const tasks = executor.tasks as ExecutionTaskInsightLocate[];

    expect(tasks.length).toBe(2);
    expect(tasks[0].status).toBe('failed');
    expect(tasks[0].error).toBeTruthy();
    expect(tasks[0].timing!.end).toBeTruthy();
    expect(tasks[1].status).toBe('cancelled');
    expect(executor.status).toBe('error');
    expect(executor.latestErrorTask()).toBeTruthy();
    expect(executor.isInErrorState()).toBeTruthy();
    expect(r).toEqual('error-output');

    // expect to throw an error
    expect(async () => {
      await executor.flush();
    }).rejects.toThrowError();

    expect(async () => {
      await executor.append(insightFindTask());
    }).rejects.toThrowError();
  });
});
