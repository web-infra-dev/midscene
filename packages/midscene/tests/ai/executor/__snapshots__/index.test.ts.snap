// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`executor > insight - basic run 1`] = `
{
  "element": {
    "center": [
      250,
      250,
    ],
    "content": "test-executor",
    "id": "0",
    "rect": {
      "height": 100,
      "left": 200,
      "top": 200,
      "width": 100,
    },
    "tap": [MockFunction spy],
  },
}
`;
